### PR TITLE
Removing Recursion from Processor

### PR DESCRIPTION
**Pull request by somedave**
This pull request requires a little explanation - sorry for the long-winded text that follows. 

Recall that I mentioned IIS was silently killing my worker process for a large-scale app on a particular complex query being run through DelegateDecompiler. I tracked the issue down to a stack overflow occurring during query execution, so I started looking for possible causes. Looking at `Processor` I found two potential sources of overflow due to recursion:

* `Processor.Process()` when called from `Processor.ConditionalBranch()` - this might be a problem with lots of nested conditionals (which my problem query has).
* `Processor.GetJoinPoint()` which uses recursion to find the common join point for nested conditional branches.

I formulated a test that uses a new thread with a limited stack size and deeply nested conditionals to trigger a stack overflow in a controlled manner, as expected. Then I removed the recursion in `Processor.Process()` by making `Processor` essentially stateless except for a single `Stack` that holds state information. Instead of recursion for conditional branches, `Processor` now iterates on the states in the `Stack`. I also used a similar strategy to remove recursion from `Processor.GetJoinPoint()` (which ended up looking a little inelegant, but it gets the job done). After applying these changes, my limited stack size tests passed.

Here's why this pull request requires explanation though - this didn't actually solve my underlying problem. Which these changes probably do reduce the likelihood of stack overflows in some situations, I discovered later that my own issue is caused by an infinite loop or deep recursion within `System.Data.Query.PlanCompiler` and `System.Data.Query.InternalTrees`. Something about my decompiled query just doesn't agree with Entity Framework, and at this point I'm more likely to think this is a bug in EF and not in DelegateDecompiler.

Regardless, the changes in this PR do result in passing the stack overflow test which failed before. I figured since the work was already done, I would see if you thought it was a good idea to merge them in.

### CODE REVIEW

File: src/DelegateDecompiler.EntityFramework.Tests/GeneratedDocumentation/DetailedListOfSupportedCommands.md [hunk 1]
```diff
@@ -1,6 +1,6 @@
 Detail of supported commands
 ============
-## Documentation produced for DelegateDecompiler, version 0.12.1.0 on Monday, 15 December 2014 02:19
+## Documentation produced for DelegateDecompiler, version 1.0.0.0 on Tuesday, 13 January 2015 14:26
```
Comment (hazzik) [inline src/DelegateDecompiler.EntityFramework.Tests/GeneratedDocumentation/DetailedListOfSupportedCommands.md]: Not related to this pull request, but we need to use AssemblyFileVersion here

**Comment by hazzik** [comment]
Thanks Dave! Can you please rebase on top of master/remove my commits from the pull request?

**Comment by somedave** [comment]
Sorry about that - I have no idea how those got in there :). I've rebased and you should be good to go. I had to force, so if you've cloned my repo or the PR repo they'll probably need to be re-cloned.

**Comment by hazzik** [comment]
Thanks.

**Comment by hazzik** [comment]
Merged, thanks

**Pull request merged**